std/typecons.d

```diff
@@ -796,35 +796,61 @@ if (distinctFieldNames!(Specs))
          *           for comparison between `Tuple`s.
          *
          * Returns:
-         * For any values `v1` on the right-hand side and `v2` on the
-         * left-hand side:
+         * For values `v1` on the left-hand side and `v2` on the right-hand side:
+         *
+         * 0 if the expression `v1 == v2` is true for all members, or else for
+         * the first member, where this is not true:
          *
          * $(UL
-         *   $(LI A negative integer if the expression `v1 < v2` is true.)
-         *   $(LI A positive integer if the expression `v1 > v2` is true.)
-         *   $(LI 0 if the expression `v1 == v2` is true.))
+         *   $(LI NaN, in case one of the operands is a NaN.)
+         *   $(LI A negative number if the expression `v1 < v2` is true.)
+         *   $(LI A positive number if the expression `v1 > v2` is true.))
          */
-        int opCmp(R)(R rhs)
+        auto opCmp(R)(R rhs)
         if (areCompatibleTuples!(typeof(this), R, "<"))
         {
             static foreach (i; 0 .. Types.length)
             {
                 if (field[i] != rhs.field[i])
                 {
+                    import std.math : isNaN;
+                    static if (isFloatingPoint!(Types[i]))
+                    {
+                        if (isNaN(field[i]))
+                            return float.nan;
+                    }
+                    static if (isFloatingPoint!(typeof(rhs.field[i])))
+                    {
+                        if (isNaN(rhs.field[i]))
+                            return float.nan;
+                    }
+
```

Comment:
Suggest adding this for user-defined types with float `opCmp`s:
```
static if (isFloatingPoint!(typeof(field[i].opCmp(field[i]))))
{
    if (field[i] != field[i])
        return float.nan;
}
static if (isFloatingPoint!(typeof(rhs.field[i].opCmp(rhs.field[i]))))
{
    if (rhs.field[i] != rhs.field[i])
        return float.nan;
}
```

Reply:
Bah. The above fails since of course `float` doesn't have an `opCmp` member. Something like this is required:
`static if (is(typeof({static assert(isFloatingPoint!(typeof(field[i].opCmp(field[i]))));})))`
For the record: that's ugly.

Reply:
What about that:

```
                    auto tmp1 = field[i] < rhs.field[i];
                    auto tmp2 = field[i] > rhs.field[i];
                    if (!tmp1 && !tmp2)
                        return float.nan;
```

Reply:
That works, but brings us back to the performance discussion from earlier.

Reply:
Also, kinda obsoletes the `isFloatingPoint` test above.

Reply:
OK. So we need something else here...

Reply:
This looks better. One very minor nitpick is it's possible to have `opCmp` return a type with an overloaded `opCmp` that returns float (and actually any number of levels to this `opCmp` tower). I'm not sure why anyone would do that, but as a serial offender on operator overload abuse, I feel it's my duty to inform people of its dangers. Should we want to actually support this, I would suggest adding this to std.traits:
``` 
template hasNanComparison(T1, T2 = T1)
{
    static if (isFloatingPoint!T1)
        enum hasNanComparison = true;
    else static if (!isAggregateType!T1)
        enum hasNanComparison = false;
    else static if (!is(typeof((T1 a, T2 b) => a.opCmp(b))))
        enum hasNanComparison = false;
    else
        enum hasNanComparison = hasNanComparison!(ReturnType!((T1 a, T2 b) => a.opCmp(b)));
}

unittest {
    assert( hasNanComparison!float);
    assert( hasNanComparison!real);
    assert(!hasNanComparison!int);
    assert(!hasNanComparison!string);
    assert(!hasNanComparison!(int*));
    assert(!hasNanComparison!(int[]));
    
    struct S
    {
        float opCmp(S s) { return float.nan; }
    }
    assert( hasNanComparison!S);
    
    struct S2
    {
        S opCmp(S2 s) { return S(); }
    }
    assert( hasNanComparison!S2);
}
```
I would argue, however, that this addition is not necessary.

Reply:
Maybe this could be done in a separate PR if anyone needs it?!? I would be happy to finish with this PR, though.

Reply:
[Agreed.](https://github.com/dlang/phobos/pull/7342)